Refactor of evented runners (#17283)

jacksontj opened the pull request:
This is a refactor of #17283 (tangentially related to #9444 since we had some discussion in there) to avoid most of the code duplication-- since we now have async client mixins.

I'm still not sure about the **progress** global, since it is _very_ specific, but I'm not sure having someone having to write out something like:

```
__jid_event__.fire_event({'data': 'Runner is {0}% done'.format(i), 'outputter': 'pprint'}, 'progress')
```

Is any better than just 

```
event.fire_event({'data': 'Runner is {0}% done'.format(i), 'outputter': 'pprint'}, 'progress', base=__tag__)
```

So in addition to the **progress** global I've added **jid**, **user**, **tag**, and **jid_event**.

As part of that the "low" func now creates a JID and events for all calls to it (so we are consistent). All of these are specific to each execution. **jid_event** is a new class called "NamespacedEvent" which is just a wrapper to allow for people to fire events from a different base tag, in this case within the runner you could fire an event with tag "foo" and it would end up as "salt/run/SOME_JID/foo" (alternatively you could just use base=**tag** with the regular event class (which I'm leaning towards, since this seems a bit messy)

@cachedout @whiteinge 
Pinging the 2 of you about the **progress** object vs the Namespaced event. I'm thinking we could maybe generalize the outputter class in the runner to work for wheel and runner, and come up with some way of printing out all events within the namespace-- so people don't have to follow the strict event format from RunnerEvent (which I've removed).

jacksontj committed:
Remove local multiprocessing, and use the aync client mixin

jacksontj committed:
Caste jid as str (its an int all over), and change runner tag to correct prefix

jacksontj committed:
Make Async really async-- and always create another process to execute the runner

jacksontj committed:
Consolidate the cmd() func to just use low(), since its all redundant (and most of the old arg parsing was redundant)

jacksontj committed:
Misc cleanup

jacksontj committed:
Move "low" to separate line-- everything else is

jacksontj committed:
Add a "namespacedevent" object, this give a convinent wrapper for firing events within a specific namespace

jacksontj committed:
Refactor-- now all the event firing is done within low()

This means that *all* runner/wheel executions will have events

jacksontj committed:
Swallow exceptions on __del__ of event

This will avoid exception in interpreter shutdown such as:

```
Exception TypeError: 'an integer is required' in <bound method MasterEvent.__del__ of <salt.utils.event.MasterEvent object at 0x3ab2a50>> ignored
```

Most of which are because during interpreter shutdown there is no order gaurantee so things (such as the zmq lib) get unloaded even though the __del__ method has references to it

jacksontj committed:
Inject globals without the sys module

jacksontj committed:
Switch to using utils.jid to generate jid

jacksontj committed:
Add a convenience wrapper for __progress__

jacksontj committed:
Remove RunnerEvent

jacksontj committed:
Fix for #18256

With this new setup of everything firing events, this seemed like a good time to tackle this infinite recursion from the reactor. What this does is pass a "user" (Reactor) to all jobs that the reactor starts. Then the reactor skips all events created by that username-- thereby only reacting to events not caused by itself.

jacksontj committed:
Pylint cleanup

jacksontj committed:
include pid in async log line

jacksontj committed:
Style cleanup

jacksontj committed:
Move get_async_returns to mixin

jacksontj committed:
Cleanup suffix determination

jacksontj committed:
Inject a "print" function into runners, and clean up (a little) the event printout handling.

jacksontj committed:
Add stdout_print test function (to test overriding print func). Also include __jid_event__.fire_event (instead of __progress__)

jacksontj committed:
add doc string

cachedout commented:
I'm fine with going the namespaced direction, so long as we update the docs and such accordingly. I think it's more flexible and all around a better approach then progress events.

jacksontj commented:
@cachedout K, I'll get a cleanup of that once I finish fixing/cleaning all the arg/kwarg parsing stuff. There are at least 3 different ways for the function calls to go and they all seem to parse it differently... so I'm just un-tangling ;)

jacksontj committed:
- Change event name for print
- include old print in new evented print
- support having the args parsed and *not* parsed in low()

cachedout commented:
Sounds great! I'm looking forward to seeing the results.

jacksontj commented:
Wheel uses "low" as kwargs, and runners seem to actually use it as a lowstate object-- so I'm going to have to make them the same-- since they have the same mixins under the hood. :/

jacksontj committed:
replace l_fun with self.function[fun], since the local no longer exists

cachedout commented:
I think that should work. I'm going to ping @whiteinge since he might want to chime in here as well.

whiteinge commented:
Agree with Thomas Jackson on namespaced direction. Also right about
lowstate diff between runner and wheel; I've had to work around that before
too.

On Wed, Dec 31, 2014, 5:28 PM Mike Place <notifications@github.com> wrote:

> I think that should work. I'm going to ping @whiteinge
> <https://github.com/whiteinge> since he might want to chime in here as
> well.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/saltstack/salt/pull/19253#issuecomment-68476535>.
>

jacksontj commented:
@whiteinge @cachedout So, I think I have it. The "lowstate" object thing that is doing is a bit odd. Instead of 'args' and "kwargs' as parts of the low object, all kwargs are basically just merged into low. I don't *think* thats what we do elsewhere, but I can't remember off hand.

jacksontj committed:
Clean up arg parsing (a little), since the low that is passed does not have a "kwargs" section

jacksontj commented:
After some quick looking that way (kwargs as part of the main dict) matches the pub data structures-- so that was probably intentional, albeit confusing.

jacksontj committed:
Add sleep to test runner

jacksontj committed:
Clean up arg parsing in runner client

whiteinge commented:
Yeah, think you're right about that. Intentional but confusing.

On Wed, Dec 31, 2014, 5:40 PM Thomas Jackson <notifications@github.com>
wrote:

> After some quick looking that way (kwargs as part of the main dict)
> matches the pub data structures-- so that was probably intentional, albeit
> confusing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/saltstack/salt/pull/19253#issuecomment-68476789>.
>

jacksontj committed:
Pass in args/kwargs

jacksontj committed:
Make the callers of this function parse their args correctly

jacksontj committed:
Consolidate more code between runner and wheel modules

jacksontj committed:
Consolidate to mixins funcs

jacksontj committed:
Consolidate to mixin funcs

jacksontj committed:
Consolidate some of the __init__  logic

jacksontj committed:
Add missing import

jacksontj committed:
Cleanup docs some

jacksontj committed:
Switch to module, instead of global (for compatibility)

jacksontj committed:
Pylint

jacksontj committed:
Remove some local files that somehow got added

jacksontj committed:
Remove local testing file

jacksontj commented:
Okay, I *think* that covers all my bases. I'll wait for the test runs to finish, but that last batch includes a *lot* of cleanup. RunnerClient and Wheel Client are basically the same-- except a couple little things. So instead of maintaining 2 completely separate classes (as we have) I've pushed even more into the mixins. This means that the interface to the wheel cmd() calls changed *slightly*, but it was kind of broken before (couldn't pass args-- for example). 

I still need to migrate the rest of the __progress__ instances to the event firing (should be simple). The only other thing I'd like to do is have the runner die if you control-c out of a synchronous call on the command line-- since that would make it backwards compatible, unless we don't care...

jacksontj committed:
Pylint cleanup

jacksontj committed:
Fix builtin access

jacksontj committed:
Cleanup __init__, since outputters are only needed in run()

jacksontj committed:
Add __init__ arg to Runner(), and call self.async (since its a subclass)

jacksontj committed:
Change how we print out events from runner/wheel calls

Now if you do something like:
```
fire_event({'data': 'Runner is {0}% done'.format(i), 'outputter': 'pprint'}, 'progress')
'Runner is 96% done'
```

```
fire_event({'data': 'Runner is {0}% done'.format(i)}, 'progress')
progress: {'data': 'Runner is 12% done'}
```

This means that you don't have to actually specify the outputters etc if you just want it to print something.

jacksontj commented:
As I'm migrating these, I wonder if we should just only print a subset of suffixes (such as new, progress, and ret) that way if you want additional events they wont *have* to be printed, and we wont have as much duplication (right now it prints tag: dict-- which is usually something like progress: data: the message here, where we could have just the dict).

jacksontj committed:
Change to new event fire_event markup

jacksontj committed:
Remove __progress__ from lint checks, and add in jid_event

jacksontj committed:
Pylint fix

whiteinge commented:
I like that idea.

On Wed, Dec 31, 2014, 7:36 PM Thomas Jackson <notifications@github.com>
wrote:

> As I'm migrating these, I wonder if we should just only print a subset of
> suffixes (such as new, progress, and ret) that way if you want additional
> events they wont *have* to be printed, and we wont have as much
> duplication (right now it prints tag: dict-- which is usually something
> like progress: data: the message here, where we could have just the dict).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/saltstack/salt/pull/19253#issuecomment-68478921>.
>

jacksontj committed:
Fix typo

jacksontj commented:
More weirdness, we actually have some tests (https://github.com/jacksontj/salt/blob/develop/tests/integration/client/runner.py#L74) which enforce the kwargs as part of the top level "low" object, just to make it more confusing

jacksontj committed:
Style cleanup

jacksontj commented:
@whiteinge which tag name should we use? Want to stick with "progress" or should we move to something more generic, like "print" or "out"? I overloaded the "print" function to fire events to the tag "print", but we can support multiple tags if we want-- or make them use the same ones.

jacksontj committed:
Add backwards compatibility to "low" function-- since the object isn't well documented we have apparently used it in all possible ways ;)

jacksontj commented:
Looks like the VMs didn't start on jenkins-- closed + re-opened to trigger a new build

jacksontj commented:
@rallytime Tests have been running for ~12 hours now, so I think its safe to say they won't ever finish ;)

jacksontj committed:
Backwards compatibility with args as well.

With format_call you can pass in positional args as kwargs with the key being the name of the variable in the function call.

jacksontj committed:
Clean up comments

jacksontj committed:
Remove PID since the process is daemonized afterwards, meaning that pid isn't correct

jacksontj commented:
@rallytime Looks like the linode account ran out of cash:

```
23:35:14 (8) Unable to add a new Linode as your account has an outstanding balance. Please submit a payment, or wait for us to automatically charge your card.
```

cachedout commented:
@jacksontj Looks like a CC expired at the end of the year. Oops! We'll get it sorted ASAP.

thatch45 commented:
Go Go Jenkins!

thatch45 commented:
Our Linode account should be fixed, I updated the card

jacksontj committed:
cleanup "Caste jid as str (its an int all over), and change runner tag to correct prefix"

This reverts most of commit 745f2d84835ae23800358cd4dcc9c8525b502714.

apparently I picked up some other changes I didn't mean to

Conflicts:
	salt/utils/event.py
	tests/integration/utils/test_reactor.py

jacksontj commented:
@rallytime Jenkins seems to be dead. I have a few open PRs and I can't seem to get it to run for any of them :(

rallytime commented:
Go Go Jenkins!

jacksontj committed:
Fix runner returns

jacksontj commented:
@rallytime @thatch45 And that they are! After the last cleanup of printing out the events I seemed to have accidentally deleted the code that set the ret. Its re-added now

thatch45 commented:
Go Go Jenkins!

thatch45 commented:
We seriously need Jenkins to be more reliable.....

jacksontj commented:
@thatch45  +1

Might I suggest by starting with some metrics/alerting tracking how its performing? Easier to fix reliability if you have metrics :)

thatch45 commented:
Looks like these failures are legit, the wheel system is closely related to runners but the test results are missing from Jenkins....

jacksontj commented:
Yea, the few that I've looked at are legitimate failures, but not legitimate tests. For example,  the integration.runners.manage.ManageTest.test_up test is calling the runner with --async and expecting the return on the CLI (d746bb2). This was because the old implementation "async" wasn't really async ;)

thatch45 commented:
Thanks!

jacksontj committed:
Add a return to the async method (jid)

jacksontj committed:
Re-add call_func backwards compat func

jacksontj commented:
I think that covers them all-- but lets wait for the test runs to finish, I don't want to accidentally break backwards compat on something ;)

jacksontj commented:
@rallytime looks like the Ubuntu VM failed immediately :(

jacksontj committed:
Pylint and style cleanup

rallytime commented:
Yes, we are having trouble with Jenkins again. We're woking on getting it more stable as we speak. :)

rallytime commented:
Go Go Jenkins!

rallytime commented:
Hrm. @jacksontj Now that we actually have the tests running here, it looks like both VMs are hanging right after this test:

test_cmd_sync (integration.client.runner.RunnerModuleTest) ... OK (0.052s)

I haven't seen this on other branches, so I think this might need to be legitimately addressed.

rallytime commented:
I ended up aborting the tests since they were hanging. Seems GH still things they're running in yellow, but they are in fact aborted where only a subset of the tests were finished before aborting.

jacksontj commented:
@rallytime I ran integration.client.runner.RunnerModuleTest (all 6 of them) locally and it passed without issue on my box :/

rallytime commented:
Ok. Jenkins is kind of backed up right now and has a pretty hefty queue. I'll circle back around to this and take another look and re-trigger these to see if it will happen again. Thanks again for your patience!

jacksontj commented:
Sure, no problem :)

jacksontj commented:
Still need to update the docs before we merge this, but I want to see if we can get a test run completed :)

jacksontj committed:
Add a blacklist of suffixes for the client mixin to print out

jacksontj commented:
Can't reproduce that test failure locally.. hopefully it was just a fluke

rallytime commented:
I haven't seen these failures on the develop branch, and we've got it pretty stable right now, so these are likely legitimate. Especially since one of the tests is failing on both test VMs. Is this an older version of develop, or has it been rebased within the last day or so? I'll see if I can reproduce these as well.

jacksontj commented:
Its a fairly old develop, so who knows. Since I was unable to repro locally I just rebased against current develop.

jacksontj commented:
@rallytime  Looks like a Jenkins rough patch... Getting all sorts of java errors :(

rallytime commented:
Go Go Jenkins!

rallytime commented:
@jacksontj These failures are consistent here, as the exact same runner tests are failing as before and they're not failing in the develop branch, so they're not left over from something else. As to why they're failing, I'll have to take a close look. :)

jacksontj committed:
Don't spin waiting for runner return, just use get_event (it has a poller)

Conflicts:
	salt/runner.py

jacksontj commented:
Those test failures should be due a change from a while ago where @cachedout had made *all* runner tests async-- which messed up some of these. I can't get the test to fail locally, but I also had some *major* rebase conflicts last attempt-- so its possible i completely messed it up ;) I just pulled in another fix for this stuff and we'll see how this test run goes.

jacksontj committed:
Revert function calls

Apparently the wheel API varies quite a bit from the runner API, we'll have to clean that up in the future

jacksontj committed:
For some reason having the event init in the shared subclass causes problems.

jacksontj committed:
Cleanup, we don't always need an "event" object, so lets not always create one

jacksontj committed:
Workaround for race condition in SaltEvent() creation and the zmq socket actually connecting

jacksontj committed:
Fix typo

jacksontj commented:
Another big rebase. So the failures weren't terribly hard to fix. The wheel ones are because the wheel interface is inconsistent with the runner ones (added TODO markers to clean it up).

The runner one was actually just exposing a race condition in the SaltEvent class. ZMQ connect() doesn't guarantee that the underlying socket is actually connected. I put in a workaround for this, but AFAIK there isn't a good fix for that.

jacksontj committed:
Add missing import

jacksontj commented:
Look like the test failures are just timeouts of tornado stuff, everything else seems fine. Those timeouts usually mean Jenkins is busy

rallytime commented:
Go Go Jenkins!

jacksontj commented:
I'd like to get this one wrapped up this week, as we *need* to get this into 2015.2, as the current runner __progress__ implementation breaks backwards compatibility and exposes an interface we don't want to support.

thatch45 commented:
Yes, we completely agree, I think that this is ready to merge

thatch45 commented:
Ok, merged!
We will need to watch the develop tests for these issues and get them cleaned up ASAP @jacksontj so that we can backport this to 2015.2

jacksontj commented:
@cachedout Would you be able to change the docs over to this new markup? I can check it after you finish-- probably a good way to get some additional code review done :)

jacksontj commented:
The only other thing I'd like to get in is some way to stop runners when they aren't run in async-- that way it would be backwards compatible.

cachedout commented:
@jacksontj Definitely. I'll do that now.

jacksontj commented:
@cachedout Any update on docs? I've actually had a couple people ping me about this :) I think I'm going to make one more change to this tomorrow to have the non-async runners execute in a thread instead of a daemonized process (so stdout print will work and so you can stop the runner execution).

cachedout commented:
@jacksontj Got distracted with some other fires. :] Will get back to docs today. Also, @kiorky has some feedback on the runner changes in #19707 if you have a moment.

s0undt3ch commented on tests/integration/utils/test_reactor.py:
```diff
@@ -0,0 +1,68 @@
+        if os.path.isdir(self.tempdir):
+            shutil.rmtree(self.tempdir)
+
+    def test_basic(self):
```
What is this supposed to test?
